### PR TITLE
Implement JoinGame Endpoint with Active Player and Queue Logic

### PR DESCRIPTION
**Pull request by riyadvaliyev**
#### **Overview**
This pull request introduces the `/joinGame` endpoint, which allows players to join either the active courts or the queue based on availability. The endpoint aligns with the latest Firestore database structure and adheres to modular and transactional principles outlined in the design specifications.

---

#### **Key Changes**

1. **Endpoint Implementation**:
   - Added the `/joinGame` endpoint to manage player placement in either active courts or the queue.
   - The endpoint checks for empty courts first and assigns the player to an available court. If no courts are available, the player is added to the queue.

2. **Transaction-Based Firestore Updates**:
   - Leveraged Firestore transactions to ensure atomic updates to the `locations` document, avoiding potential collisions with other endpoints.
   - Updated fields:
     - `activeFirebaseUIDs`
     - `activeNicknames`
     - `activeStartTimes`
     - `activeWaitingPlayers`
     - `queueFirebaseUIDs`
     - `queueNicknames`

3. **Input Validation**:
   - Ensures that all required fields (`location`, `firebaseUID`, `nickname`) are present and valid.

4. **Response Messaging**:
   - Returns clear responses to the frontend indicating whether the player was added to the active courts or the queue.

5. **Testing & Debugging**:
   - Thoroughly tested the endpoint using Postman to ensure:
     - Players are correctly assigned to active courts when available.
     - Players are added to the queue when no courts are available.
     - Proper handling of invalid input and edge cases (e.g., queue full).
   - Included logs to confirm the correct state of Firestore data during updates.

---

#### **Testing Evidence**
1. **Test Case: Adding a Player to an Empty Court**:
  
Screenshot 2024-11-21 at 10 47 23 PM
Screenshot 2024-11-21 at 10 48 22 PM

2. **Test Case: Adding a Player to the Queue**:
Screenshot 2024-11-21 at 10 47 47 PM
Screenshot 2024-11-21 at 10 48 09 PM

---

#### **How to Test**
1. Use Postman or a similar tool to send POST requests to the `/joinGame` endpoint.
2. Ensure the following scenarios work as expected:
   - Player is added to active courts when there is availability.
   - Player is added to the queue when all courts are occupied.
   - Proper error handling for invalid inputs.
3. Check Firestore to verify updates to the `locations` document.

---

#### **Next Steps**
- A separate PR will address the `dynamicBuffer` function, which integrates with this endpoint to manage active player timing and queue advancement logic.

---

This implementation resolves the functionality for managing player placement, adheres to modular design principles, and ensures efficient Firestore usage through transactions. Let me know if further refinements or additional tests are required!

### CODE REVIEW

File: backend/utils/joinGame.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
+const admin = require('firebase-admin');
+
+async function joinGame(locationData, firebaseUID, nickname) {
+    const {
+        activeFirebaseUIDs,
+        activeNicknames,
+        activeStartTimes,
+        activeWaitingPlayers,
+        queueFirebaseUIDs,
+        queueNicknames,
```
Comment (Ron-Madan) [inline backend/utils/joinGame.js:10]: Please include queueJoinTimes as well as we now need to store the queue join time

File: backend/utils/joinGame.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        if (queueFirebaseUIDs.length >= 5) {
+            return {
+                success: false,
+                message: 'Queue is full. Cannot join.',
+            };
+        }
+
+        // Add the player to the queue
+        queueFirebaseUIDs.push(firebaseUID);
+        queueNicknames.push(nickname);
```
Comment (Ron-Madan) [inline backend/utils/joinGame.js:52]: Add the timestamp to the queueJoinTimes array

**Review by Ron-Madan — APPROVED** [review]
LGTM!
